waku/node/waku_node.nim

```diff
@@ -482,35 +581,86 @@ proc filterUnsubscribe*(node: WakuNode, pubsubTopic: Option[PubsubTopic], conten
-proc subscribe*(node: WakuNode, pubsubTopic: Option[PubsubTopic], contentTopics: ContentTopic|seq[ContentTopic], handler: FilterPushHandler) {.async, gcsafe,
-  deprecated: "Use the explicit destination peer procedure. Use 'node.filterSubscribe()' instead.".} =
-  ## Registers for messages that match a specific filter. Triggers the handler whenever a message is received.
-  if node.wakuFilterClientLegacy.isNil():
-    error "cannot register filter subscription to topic", error="waku filter client is nil"
-    return
+proc filterUnsubscribe*(node: WakuNode,
+                          pubsubTopic: Option[PubsubTopic],
+                          contentTopics: Option[seq[ContentTopic]],
+                          peer: RemotePeerInfo|string):
 
-  let peerOpt = node.peerManager.selectPeer(WakuFilterCodec)
-  if peerOpt.isNone():
-    error "cannot register filter subscription to topic", error="no suitable remote peers"
-    return
+                Future[FilterSubscribeResult]
 
-  await node.filterSubscribe(pubsubTopic, contentTopics, handler, peer=peerOpt.get())
+                {.async, gcsafe, raises: [Defect, ValueError].} =
 
-# TODO: Move to application module (e.g., wakunode2.nim)
-proc unsubscribe*(node: WakuNode, pubsubTopic: Option[PubsubTopic], contentTopics: ContentTopic|seq[ContentTopic]) {.async, gcsafe,
-  deprecated: "Use the explicit destination peer procedure. Use 'node.filterUnsusbscribe()' instead.".} =
-  ## Unsubscribe from a content filter.
+  ## Unsubscribe from a content filter V2".
   if node.wakuFilterClientLegacy.isNil():
     error "cannot unregister filter subscription to content", error="waku filter client is nil"
-    return
+    return err(FilterSubscribeError.serviceUnavailable())
 
-  let peerOpt = node.peerManager.selectPeer(WakuFilterCodec)
-  if peerOpt.isNone():
-    error "cannot register filter subscription to topic", error="no suitable remote peers"
-    return
+  let remotePeerRes = parsePeerInfo(peer)
+  if remotePeerRes.isErr():
+    error "couldn't parse remotePeerInfo", error = remotePeerRes.error
+    return err(FilterSubscribeError.serviceUnavailable("No peers available"))
+
+  let remotePeer = remotePeerRes.value
+
+  if pubsubTopic.isSome():
+    if contentTopics.isSome():
+      info "deregistering filter subscription to content", pubsubTopic=pubsubTopic.get(), contentTopics=contentTopics.get(), peer=remotePeer.peerId
+
+      let unsubRes = await node.wakuFilterClient.unsubscribe(remotePeer, pubsubTopic.get(), contentTopics.get())
+      if unsubRes.isOk():
+        info "unsubscribed from topic", pubsubTopic=pubsubTopic.get(), contentTopics=contentTopics.get()
+      else:
+        error "failed filter unsubscription", error=unsubRes.error
+        waku_node_errors.inc(labelValues = ["unsubscribe_filter_failure"])
+
+      return unsubRes
+
+  elif contentTopics.isNone(): # none of pubsubTopic and contentTopics are not provided, go for unsub all
```

Comment:
This logic should not exist in the `waku_node`. It simply passes the `pubsubTopic` and `contentTopics` to the underlying client. The client has the mandate to implement the protocol here, which states that and empty `contentTopics` is an error. In fact, this would not even allow the possibility for a client to `unsubscribeAll` for all `pubsubTopics`. The node simply exposes this underlying functionality in an integrated API (in other words, should add a `filterUnsubscribeAll*` method). Even the fact that the node is applying autosharding rules here should probably be abstracted away, but that is too much for this PR.